### ISSUE TITLE
Store conflicts between root trees

### ISSUE THREAD

**Issue by martinvonz** [martinvonz/jj#1624]
This is a tracking issue for moving conflict storage from paths (stored in trees) to the root tree (stored in commits). Currently, when we run into a conflict that we can't automatically resolve, we store it as a conflict associated with the path. The proposal is the instead bubble the conflict up so it becomes a conflict between root tree objects, which we'd store in the commit object. We already bubble up conflicts from the hunk level to the file level, so it's more consistent to bubble it up all the way to the root tree. This was originally proposed by @ilyagr in https://github.com/martinvonz/jj/pull/1480#issuecomment-1498249515. 

This proposal makes it really fast to find out whether a revision is conflicted (you can tell from the commit object). This is very useful for use in the `jj log` output. It's also reasonably fast to find out which paths are conflicted  by performing the merge of the trees recorded in the commit object. That's useful for `jj status`. Before this proposal, we had planned an on-disk cache of which trees had a conflict. This proposal obviates the need for that and results in a simpler model.

With conflicts stored in the root tree, we have to do a full tree merge of the root trees every time we need materialized conflicts, which we do e.g. when we check out or diff with a conflicted commit.

Rough plan (many of the steps are independent):

1. Make tree-merge code able to merge an arbitrary number of trees
2. Add another version of the tree-merge function, which materializes conflicts and returns the tree (without any conflicts in it) and a list of conflicts objects
3. Add yet another version of the tree-merge function, which bubbles conflicts up to the root tree (as above, no conflicts in the individual trees)
4. Update call sites to read and write conflicts from either paths or root tree
5. Update `working_copy` to not store conflict state but to get the conflict from the merged base trees
6. (Much later: ) Drop support for reading and writing conflicts from the backend

I'm sure I'll discover other necessary steps along the way. PR #1610 is my first PR towards getting this done.

I plan to add a config option to control the release of this change. Once the code is in place, we will always read conflicts from the commit object when we see it, and there would be an option to also scan the whole tree for conflicts as we currently do. There will also be an option to not scan the tree (useful in new repos, especially large ones). It should also be possible to set the config so tree conflicts are not written. That will be the default value for a while until we're confident in the new model.

**Comment by martinvonz** [comment]
Here's an update after having thought some more about this. Some of this is just restructured from the message above.

We interact with conflicts in different ways in different context. Here is how how we interact with them in each context, and the requirements they impose:

* Log: Indicate commits with any conflicts. This is currently done by walking the whole tree. With tree-level conflicts, we have the the information in the commit object.

* Status: Indicate which paths have conflicts. This is currently also done by walking the whole tree. With tree-level conflicts, we will be able to skip sub-trees that can be trivially resolved.

* Update working copy: When checking out a tree into the working copy, we materialize the conflicts are conflict markers. With tree-level conflicts, we need to be able to be able to do the same. That's what step 2 above was about. I'm now thinking that we should instead preserve the tree-level conflict so we can use it when snapshotting.

* Snapshot working copy: When a conflict has been edited in the working copy, we parse the conflict markers and build up a new conflict (or treat it as resolved if there are no conflict markers). Then we write the new conflict's id to the `TreeBuilder`. With tree-level conflicts, we would need to build 3 new trees instead of one (in the 3-way case).

* Diff: We currently materialize the conflicts the same way we do in the working copy, but since we have the conflict objects, we have the opportunity to render diffs involving conflicts differently (e.g. like `git diff --cc`). I would like to keep that possibility, which means that it would be good if we could somehow extract path-level conflict objects given a tree-level conflict, even if the plan is to not persist those path-level conflict objects.

* Merge: We currently write conflict objects to the backend, and then record the conflict id in the tree object. With tree-level conflicts, we want to bubble those up to the root. That's what step 3 above was about. However, what I think I had not thought about earlier was that we also want to be able to merge conflicted trees with other conflicted trees.

My current thinking is that we should create a new `MergedTree` type representing the possibly conflicted tree. It might look like this:

```
enum MergedTree {
  Resolved(Tree),
  Conflicted {
    removes: Vec<Tree>,
    adds: Vec<Tree>,
  }
}
```

I'm thinking it would have iterators and methods for accessing the value at a specific path, which would work by doing the merge on the fly. That would be unlike how we currently materialize trees by writing back merged subtrees to the backend. Then we can define diff and merge algorithm on the `MergedTree` type. I'll start working this soon, so we'll see how it turns out.

**Comment by ilyagr** [comment]
This is a minor thing, but is the `Resolved` arm of the `MergedTree` enum necessary? I think we could just have a conflict with a single add and no removes instead.

I'm also wondering if there's a way to make sure that `# of adds = 1 + (# of removes)` in that type.

Re snapshotting and materializing the working copy, here's a detail that mildly bothers me, though I don't think it's a blocker. Let's say we did a rebase that created a 3-way conflict in a file `a`. Let's say we then did another rebase of the result that created a 3-way conflict in another file `b` as well. The *tree* conflict will now have more than 3 sides (4 adds and 3 removes, I think). In present-day jj, however, each file would be materialized as a 3-way conflict.

I guess, after we change to tree-level conflicts, we could at first we could materialize each file as a 4-way conflict. It might be nicer to materialize them as 3-way conflicts, but for the snapshotting to work, that would require labeling which sides of the 4-way conflict each side corresponds to.

**Comment by martinvonz** [comment]
> This is a minor thing, but is the `Resolved` arm of the `MergedTree` enum necessary? I think we could just have a conflict with a single add and no removes instead.

We may still want the `Resolved` variant for the fast path. I guess I'll see when I start working on it (i.e. very soon).

> I'm also wondering if there's a way to make sure that `# of adds = 1 + (# of removes)` in that type.

We could do `base: Tree, diffs: Vec<(Tree, Tree)>`, but I'm not sure it's worth it because I think that will make it harder to work with.

> Re snapshotting and materializing the working copy, here's a detail that mildly bothers me, though I don't think it's a blocker. Let's say we did a rebase that created a 3-way conflict in a file `a`. Let's say we then did another rebase of the result that created a 3-way conflict in another file `b` as well. The _tree_ conflict will now have more than 3 sides (4 adds and 3 removes, I think). In present-day jj, however, each file would be materialized as a 3-way conflict.

Yeah, I was also thinking about that. I think we can do the simplification step before materializing and then when we parse the conflict markers, we should be able to map each hunk back to the right term in the unsimplified conflict. As you said, we might just start with presenting the full conflict, even if that means that some diffs are empty in the conflict marker.

**Comment by martinvonz** [comment]
Here's something I had somehow not considered before: writing conflicts to the working copy and reading them back again is lossy because we resolve trivial conflicts before we write out the conflict markers. Let's say we have a 3-way conflict in a file with these states:

Base:
```
a-base
b-base
c-base
```

Left:
```
a-left
b-base
c-left
```

Right:
```
a-base
b-base
c-right
```

When we write the conflict markers for that, it's going to look like this:
```
a-left
b-base
<<<<<<<
+++++++
c-left
%%%%%%%
-c-base
+c-right
>>>>>>>
```

The key is that the first hunk is trivially resolved as "a2". When we read back the conflict from the working copy, we're going to get a conflict between these states:

Base:
```
a-left
b-base
c-base
```

Left:
```
a-left
b-base
c-left
```

Right:
```
a-left
b-base
c-right
```

That's a different conflict than what we had written. There are actually two reasons we won't read back an unchanged conflict. The first is that the mtime is usually unchanged, and even if it's not, we're going to [return early when the content is unchanged](https://github.com/martinvonz/jj/blob/ca6b9828d1e87585b63e6550331541159d7727d3/lib/src/conflicts.rs#L222-L224). Still, this makes me think that we should change the behavior so the roundtrip via the working copy doesn't change the conflict even if we didn't have those checks preventing it right now. We can do that by resolving the trivial hunks already while merging the trees, so the conflict exactly matches the conflict markers we'd write. Another option is to make sure that we can read back the same conflict from the working copy. We could do that by writing conflict markers even for hunk that can be trivially resolved, but that would be confusing for the user. It would be better to instead check when we read back the contents if some parts match the trivial resolution of a hunk, just like we already do at the file level (the code I linked to above).

The reason I started thinking about this now is for tree-level conflicts (i.e. the bug I'm commenting on). I think the problem would become more obvious there, because snapshotting the working copy when there are conflicts is likely to result in some trees to be updated. For example, if one side of the conflict modified file A and B and the other side modified only A (in a way that conflicted), then the next snapshot would propagate the state of the modified B file to the other side and the base.

Between the two main solutions, I think we should go with the first one, i.e. to resolve trivial conflicts also at the hunk level when we merge the trees. I think that will be much simpler, especially when considering tree-level conflicts. It also means less repeated work to re-resolve conflicts when we access a commit with conflicts. We can even simplify the algorithm for listing conflicts if we can trust that any conflict in a `FileId` is a conflict (without the change proposed here, we have to redo the content merge to see if it's actually a conflict).

So, unless I hear any reasons not to, I'm going to start by changing our current tree-merge algorithm to try to partially resolve any content conflicts (i.e. when some hunks can be trivially resolved).

**Comment by ilyagr** [comment]
> Between the two main solutions, I think we should go with the first one, i.e. to resolve trivial conflicts also at the hunk level when we merge the trees.

There are things to worry about here. For example, the file could be structured data (say JSON) that the user was going to `jj resolve` with a custom merge tool. Auto-merge could mess up the syntax.

I don't know how serious or important this is, but it would be nice if `jj` preserved enough information to do that correctly, at least in principle.

E.g., in  your example `a-left` could contain more `{`-s than `a-base`, and there could be an extra closing bracket buried in some conflict way below.

**Comment by martinvonz** [comment]
> > Between the two main solutions, I think we should go with the first one, i.e. to resolve trivial conflicts also at the hunk level when we merge the trees.
> 
> There are things to worry about here. For example, the file could be structured data (say JSON) that the user was going to `jj resolve` with a custom merge tool. Auto-merge could mess up the syntax.
> 
> I don't know how serious or important this is, but it would be nice if `jj` preserved enough information to do that correctly, at least in principle.
> 
> E.g., in your example `a-left` could contain more `{`-s than `a-base`, and there could be an extra closing bracket buried in some conflict way below that's not auto-resolvable. Then, the external merge tool would get invalid JSON on at least one side.

Ah, yes. I'm not sure what to do about that. I suppose one option would be to add configuration (in the tree or outside) that says that jj should not attempt to resolve conflicts in some files. I think there are also cases where it's useful to do the partial conflict resolution, because it can reduce the total number of conflicts we leave for the user when a commit has been rebased several times. But that's just my feeling, I haven't worked through any actual examples where it would make a difference.

By the way, a similar problem can also occur at the tree level. For example, it might be that one file should always changed when another file was changed (e.g. a `.rs` file compiled from a `.proto` file). I think that's what Mercurial has (had?) a "merge driver" feature for - for allowing an external tool to control the merge process.

For now, I'll at least make it so we resolve file-level conflicts when using tree-level conflicts, instead of leaving the whole trees unchanged.

**Comment by ilyagr** [comment]
I have some thoughts, but they are likely not actionable.

> By the way, a similar problem can also occur at the tree level. For example, it might be that one file should always changed when another file was changed (e.g. a .rs file compiled from a .proto file). I think that's what Mercurial has (had?) a "merge driver" feature for - for allowing an external tool to control the merge process.

For this reason, my original idea was to store the exact trees for every side of the conflict unchanged, without simplification. This clearly causes some problems (e.g. how to show diffs), but maybe we could figure out a way past them. Ideally, everything would "just work" and there was a command that would change the way the conflict is represented on disk.

> I think there are also cases where it's useful to do the partial conflict resolution, because it can reduce the total number of conflicts we leave for the user when a commit has been rebased several times. But that's just my feeling,

I also have this feeling. If we had a perfectly consistent theory of rebasing conflicts, this shouldn't happen, but we know our theory isn't perfectly consistent (order of rebasing matters).

**Comment by ilyagr** [comment]
My first instinct would be to make everything configurable: specify the degree of simplification of file conflicts per file (with something like `.gitattributes` or a jj version of it).

Trees could be simplified before rebase (with a config option to not do that).

I'm not sure whether this is wise in terms of usefulness/complexity ratio.

**Comment by martinvonz** [comment]
> I also have this feeling. If we had a perfectly consistent theory of rebasing & simplifying conflicts, this shouldn't happen, but we know our theory isn't perfectly consistent (order of rebasing matters). This is a problem for achieving the ideal I described in my previous paragraph.

Yes, the "both sides made the same change" rule (`A+A-B=>A`) makes it not perfect (I think that's what you also meant by "order of rebasing matters"). Since we have that rule, and we know it means that rebasing via an intermediate commit might give different results, I think it's actually going to be less surprising to resolve any conflicts we can instead of leaving the trees completely unchanged when there's a conflict. Let's say you have this history:

```
C Y
| |
B X
|/
A
```

If we now rebase X and Y onto B and then onto C, we can get different results because of the "same change" rule. A typical example is when B equals X and C undoes those changes (i.e. its state is equal to A's). If we rebase directly to C, X and Y would have the same diff (and content) as before the rebase. If we rebase via B, X would become empty, and would of course remain empty after the rebase onto C.

Now, let's say we added some changes to B and X, such that they conflicted. Let's say C still undoes B. Rebasing directly to C will then preserve X and Y, as before. However, the intermediate result on top of B is going to have a conflict, so we leave the conflict in the commit. Then we rebase it to C and the conflict goes away, and X and Y will be unchanged just like they were when we rebased directly to C.

I think it's surprising that the unrelated conflict we added makes the result so different. To clarify, the above behavior is how jj currently behaves if a single file was changes in these commits and the additional conflict was added somewhere else in the file. With tree-level conflicts, we'd get that behavior if the conflict was added anywhere in the tree.

As we have talked about before, one solution is to drop the "same change" rule, but we decided that that is less user-friendly in most cases, because the user will almost always want to resolve it to what the two sides agreed on. So, if we keep the rule, I think we should instead resolve as much as possible when we merge two trees. That includes resolving down to the hunk level (we currently stop at the file level). That would remove the impact of the additional conflict on the rest of the rebase. It would not help with the inconsistent result between the direct rebase to C and the indirect rebase via B.

Or maybe this issue is another reason to reconsider the "same change" rule. We would do that by addressing [the TODO about making `trivial_merge()` more strict](https://github.com/martinvonz/jj/blob/23351c32eaeb728830c57205d063bd0ea237be60/lib/src/merge.rs#L74-L76). Maybe the UX wouldn't be so bad if we added e.g. `jj resolve --same-change`?

By the way, another change we could make is to make `rebase` (and related commands) effectively rebase one commit at a time (first to B, then to C in the example above). We could skip commits that touch non-overlapping sets of files to make it more efficient. It would probably be fast enough even in Google's repo, because most rebases don't touch the same files much, and we have an index from path to commits that touch the path. If we did that, then we would make the direct and indirect rebases behave the same, whether or not we drop the "same change" rule.

I should also mention that I think I've seen the example above being used as an argument against 3-way merging. I think the argument was that if we merge C and Y there, we should notice that B is the same patch as X, so we should skip X instead of applying X again onto C. The reasoning is that maybe X was a cherry-pick to a release branch and C was a backout/revert of B because it was no longer needed on the main branch. I think patch-based VCSs handle that by considering B and X the same patch (even if some conflicts needed to be resolved?), but snapshot-based VCSs will have to use heuristics or some other tracking of it. Still, I think all of the above applies equally if we imagine that X has a subset of the changes in B instead of having exactly the same changes.

**Comment by ilyagr** [comment]
I haven't worked through most of your comment, but I have a bit of an aside.

>Yes, the "both sides made the same change" rule (A+A-B=>A) makes it not perfect (I think that's what you also meant by "order of rebasing matters")

That is what I originally meant, but on second thought there are more reasons for it. I think the way the diff algorithm splits file into hunks might be different depending on order of rebases, resulting in different conflicts. Also, I never quite worked through it, but there is an example in https://pijul.org/manual/why_pijul.html#associativity

**Comment by martinvonz** [comment]
I think we can consider this done now. We just need to experiment with it for a while and then enable it by defualt.
